protobuf/plugin/deepcompare/deepcompare.go

```diff
@@ -0,0 +1,293 @@
+	d.Out()
+	d.P("}")
+
+	oneofByIndex := [][]*descriptor.FieldDescriptorProto{}
+	for _, f := range m.Field {
+		fName := generator.CamelCase(*f.Name)
+		if gogoproto.IsCustomName(f) {
+			fName = gogoproto.GetCustomName(f)
+		}
+
+		// Handle oneof type, we defer them to a loop below
+		if f.OneofIndex != nil {
+			if len(oneofByIndex) <= int(*f.OneofIndex) {
+				oneofByIndex = append(oneofByIndex, []*descriptor.FieldDescriptorProto{})
+			}
+
+			oneofByIndex[*f.OneofIndex] = append(oneofByIndex[*f.OneofIndex], f)
+			continue
+		}
+
+		// Handle all kinds of message type
+		if f.IsMessage() {
+			// Handle map type
+			if d.genMap(m, f) {
+				continue
+			}
+
+			// Handle any message which is not repeated or part of oneof
+			if !f.IsRepeated() && f.OneofIndex == nil {
+				if !gogoproto.IsNullable(f) {
+					d.genEqualFunc("&m."+fName, "&o."+fName)
+				} else {
+					d.genEqualFunc("m."+fName, "o."+fName)
+				}
+				continue
+			}
+		}
+
+		// Handle repeated field
+		if f.IsRepeated() {
+			d.genRepeated(m, f)
+			continue
+		}
+
+		// Handle bytes
+		if f.IsBytes() {
+			d.genBytesEqual("m."+fName, "o."+fName)
+			continue
+		}
+
+		// scalar: normal equality
+		d.genScalarEqual("m."+fName, "o."+fName)
+	}
+
+	for i, oo := range m.GetOneofDecl() {
+		d.genOneOf(m, oo, oneofByIndex[i])
+	}
+
+	d.P("return true")
+	d.Out()
+	d.P("}")
+	d.P()
+}
+
+func (d *deepCompareGen) genMap(m *generator.Descriptor, f *descriptor.FieldDescriptorProto) bool {
+	fName := generator.CamelCase(*f.Name)
+	if gogoproto.IsCustomName(f) {
+		fName = gogoproto.GetCustomName(f)
+	}
+
+	dv := d.ObjectNamed(f.GetTypeName())
+	desc, ok := dv.(*generator.Descriptor)
+	if !ok || !desc.GetOptions().GetMapEntry() {
+		return false
+	}
+
+	mt := d.GoMapType(desc, f)
+
+	d.genScalarEqual("len(m."+fName+")", "len(o."+fName+")")
+	d.P("for k, vo := range o.", fName, " {")
+	d.In()
+	d.P("vm, ok := m.", fName, "[k]")
+	d.P("if !ok {")
+	d.In()
+	d.P("return false")
+	d.Out()
+	d.P("}")
+	if mt.ValueField.IsMessage() {
+		if !gogoproto.IsNullable(f) {
+			d.genEqualFunc("&vo", "&vm")
+		} else {
+			d.genEqualFunc("vo", "vm")
+		}
+	} else if mt.ValueField.IsBytes() {
+		d.genBytesEqual("vo", "vm")
+	} else {
+		d.genScalarEqual("vo", "vm")
+	}
+	d.Out()
+	d.P("}")
+	d.P()
+
+	return true
+}
+
+func (d *deepCompareGen) genRepeated(m *generator.Descriptor, f *descriptor.FieldDescriptorProto) {
+	fName := generator.CamelCase(*f.Name)
+	if gogoproto.IsCustomName(f) {
+		fName = gogoproto.GetCustomName(f)
+	}
+
+	d.genScalarEqual("len(m."+fName+")", "len(o."+fName+")")
+
+	if proto.GetBoolExtension(f.Options, plugin.E_OrderInsensitive, false) {
+		// Order insensitive compare
+		d.P("remaining", fName, " := make(map[int]struct{})")
+		d.P("for i := range o.", fName, " {")
+		d.In()
+		d.P("remaining", fName, "[i] = struct{}{}")
+		d.Out()
+		d.P("}")
+		d.P("outer", fName, " :")
+		d.P("for i := range m.", fName, " {")
+		d.In()
+		d.P("for j := range remaining", fName, " {")
```

Comment:
> I was thinking recursive with hashcode like this.

It looks like that involves hashing. It's not easy to hash arbitrary protobuf messages. They can contain many different primitive data types, which we would need to handle individually. There are some special cases that are especially difficult, like hashing maps - you need to sort the keys to make sure the hash comes out the same between two computations.

Essentially, you need to write a serializer for protobuf, and then hash the serialized message bytes. Unfortunately the built-in serialization isn't good enough, because it writes maps in random order.

So it's not impossible to do this, but it's a lot of code. I think if we must have a non-quadratic algorithm, it's a lot simpler to change the `Equal` method to `Compare` method that returns a number less than, equal to, or greater than 0. Then we could use that to sort both repeated fields out-of-place, and then compare them directly. I think this is much simpler than hashing. I can do this if you think it's worth it. It's probably a few hours of work. It also has the advantage that it won't be confused with the built-in gogo plugin that generates boolean-valued `Equal` methods.

On the other hand, it will probably be slower in practice for the kind of input sizes we're talking about.

> I agree. But at the same time lower level code shouldn't make assumption on high level usage.

Yes, it's hard to argue with this. I didn't want to make this code too complicated with optimizations that aren't necessary right now, but I also don't like introducing quadratic behavior that might cause problems later. If you want, I can change this to use the sort-based algorithm I described above. Or if you have any ideas that don't involve hashing, I can consider those.